### ISSUE TITLE
Get rid of Http2StreamRemovalPolicy

### ISSUE THREAD

**Issue by nmittler** [netty/netty#3448]
The `DefaultHttp2StreamRemovalPolicy` by default waits 5 seconds after a stream has been closed before removing it from the connection.  The reason for this is to support the following case from the HTTP2 spec (From [Section 5.1](https://tools.ietf.org/html/draft-ietf-httpbis-http2-17#section-5.1)):

    "WINDOW_UPDATE or RST_STREAM frames can be received in this state
     for a short period after a DATA or HEADERS frame containing an
     END_STREAM flag is sent.  Until the remote peer receives and
     processes RST_STREAM or the frame bearing the END_STREAM flag, it
     might send frames of these types.  Endpoints MUST ignore
     WINDOW_UPDATE or RST_STREAM frames received in this state, though
     endpoints MAY choose to treat frames that arrive a significant
     time after sending END_STREAM as a connection error
     of type PROTOCOL_ERROR."

These cases don't really seem to warrant the extra complexity that DefaultHttp2StreamRemovalPolicy introduces.  It would be better to just always ignore WINDOW_UPDATE and RST_STREAM for non-existent streams.  This way we can always remove the streams as soon as they're closed to avoid potential memory leaks.

**Comment by nmittler** [comment]
@Scottmitch WDYT?

**Comment by jpinner** [comment]
We ignore the frames for non-existent streams if they are for stream Ids below the lastGoodStreamId

**Comment by louiscryan** [comment]
That seems like a reasonable default policy. 

I guess the question here is whether we need to continue to allow for a pluggable policy or just bake this behavior in.

**Comment by Scottmitch** [comment]
@jpinner - What is `lastGoodStreamId`?  Does this relate to the stream id from a [GOAWAY](https://tools.ietf.org/html/draft-ietf-httpbis-http2-17#section-6.8)?

@nmittler - I would like to understand the implications of this statement:

>This way we can always remove the streams as soon as they're closed to avoid potential memory leaks.

Does `DefaultHttp2StreamRemovalPolicy` itself have a problem, or the way the interfaces are designed make memory leaks likely?

My first thought are at a minimum we should do some re-naming because the [immediateRemovalPolicy](https://github.com/netty/netty/blob/master/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2CodecUtil.java#L119) is really the [default](https://github.com/netty/netty/blob/master/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java#L69).  My next thoughts are I have not yet had a need for `DefaultHttp2StreamRemovalPolicy` or to use anything but the  [immediateRemovalPolicy](https://github.com/netty/netty/blob/master/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2CodecUtil.java#L119).  I don't necessarily have a problem with the `Http2StreamRemovalPolicy` interface but if we are trying to simplify perhaps a better extension point would be to make [DefaultHttp2Connection.deactivateInternal](https://github.com/netty/netty/blob/master/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java#L226)'s call to `removalPolicy.markForRemoval(stream);` a bit more accessible?  Something like adding a method to DefaultHttp2Connection that is called instead of using the `removalPolicy`:
```java
protected void markForRemoval(Http2Stream stream) {
  removeStream(stream);
}
```

We could then get rid of `Http2StreamRemovalPolicy`.  Less object creation, conditional checking, method call redirection, garbage creation, and less indirection.  WDYT?

**Comment by jpinner** [comment]
@Scottmitch sorry, should have been more explicit -- yes, it's the stream id that would go in a goaway frame if you were to start shutting down the connection

**Comment by Scottmitch** [comment]
@nmittler - I re-read your issue description and looked at the interfaces a bit more...I think I came to the same conclusion as you did :)

Just call [DefaultHttp2Connection.removeStream](https://github.com/netty/netty/blob/master/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java#L192) from [DefaultHttp2Connection.deactivateInternal](https://github.com/netty/netty/blob/master/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java#L226).  Consider making [DefaultHttp2Connection.removeStream](https://github.com/netty/netty/blob/master/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java#L192) `protected` if we want to preserve the "removal" hooks.  Is this what you were originally suggesting?

**Comment by Scottmitch** [comment]
@jpinner - Thanks for clarifying, and this makes sense.

@nmittler - Could we just use (https://github.com/netty/netty/blob/master/codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java#L377) for this purpose.  I'm hesitant because the application has the ability to send a go away with what ever stream id they want.  We may need to consider providing an extension point which returns `connection.remote().lastStreamCreated()` by default.  This would give the application the opportunity to override our somewhat speculative question "if we were to send a go away now, what stream id would the application want to use".

**Comment by nmittler** [comment]
@Scottmitch I think we should just get away of everything related to stream "removal" and just have "closed" be the end state.

I think for now we can just use `connection.remove().lastStreamCreated()` ... this would maintain the current behavior.  We can look for a way of allowing the application to override this in the future, but I think that's beyond the scope here.  WDYT?

**Comment by Scottmitch** [comment]
@nmittler - I think we are mostly in agreement and we can work out the details in the PR.

> I think for now we can just use connection.remove().lastStreamCreated() ... this would maintain the current behavior. 

This is not obvious to me yet. [onWindowUpdateRead](https://github.com/netty/netty/blob/master/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java#L531) and [onRstStreamRead](https://github.com/netty/netty/blob/master/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java#L388) both use `connection.requireStream(...)`.  This will change to support the ignore "frames for streams that don't exist but are lower than the stream id that would be sent in a go away frame".  The user currently has some control over when errors get returned in these situations via the "removal".  If we force `connection.remote().lastStreamCreated()` we may be taking this control away.  I think this is a good default but the question is should we be providing an extension point as we introduce these changes.

**Comment by nmittler** [comment]
@Scottmitch agreed, that's a good point.  Do you have any thoughts on what the API might look like?

**Comment by Scottmitch** [comment]
Hmm I can't really dig too deep yet but here is an option:

Add something like `int lastGoodStream()` to `Http2Connection.EndPoint` or `Http2Connection`. Then either:
1. just return `connection.remote().lastKnownStream()` from the default classes (extension point is  extend the class and override method). OR;
2. create an interface which can be extended (see below) and call this interface to do the work for you.

```java
public interface Http2StreamStatus {
	public static final Http2RemoteStreamStatus DEFAULT = new Http2RemoteStreamStatus() {
		@Override
		public int lastGoodStreamId(Http2Connection connection) {
			return connection.remote().lastKnownStream();
		}
	};

	int lastGoodStreamId(Http2Connection connection);
}
```

Thoughts? Can you come up with anything better?

**Comment by Scottmitch** [comment]
We should also consider how to reconcile any potentially new interfaces with the stream id parameter on `Http2LifecycleManager.writeGoAway`, `Http2ConnectionHandler.writeGoAway`, `DefaultHttp2ConnectionEncoder.writeGoAway`, etc... Should we not allow a stream ID here, and just use the stream ID from this new interface?  Is there a more straightforward way to present this?

**Comment by nmittler** [comment]
That seems reasonable.  I might suggest some name changes (this can probably be improved as well):

```java
public interface Http2GoAwayProvider {
	public static final Http2GoAwayProvider DEFAULT = new Http2GoAwayProvider() {
		@Override
		public int goAwayStream(Http2Connection connection) {
			return connection.remote().lastKnownStream();
		}
	};

	int goAwayStream(Http2Connection connection);
}
```

**Comment by Scottmitch** [comment]
@nmittler - SGTM.  Should have added a caveat: names are for demo purposes only...use at your own risk ;)

**Comment by nmittler** [comment]
Ha! We should have that disclaimer on all of our .java files :)

**Comment by nmittler** [comment]
@Scottmitch I'm just getting back to this.  I think in general, we can get away with removing streams as we close them. In all the inbound frame handler methods, we could more-or-less ignore frames for streams that have been closed (i.e. existing-and-closed or the streamId <= `lastKnownStream` created by either `Endpoint`).

We could be more picky and maintain some history of streamIds for which we've sent `RST_STREAM` in either direction and trigger errors if an `Endpoint` sends a (disallowed) frame after it had already sent `RST_STREAM`.  However, I'm not sure it would be worth it - it might be simpler and with less overhead to just to ignore all frames (to the extent possible) for any historical stream (existing-and-closed or possibly-existed).

The only caveat to this (that I can think of ATM) would be the priority tree for outbound flow control.  I believe we need to keep any closed streams around so long as they still have children.  Once all of their children are removed, the closed stream can then be removed from the connection entirely. Actually now that I'm thinking about it, this might be a separate task unto itself since this would currently be a problem if using the default (immediate) stream removal policy.  We should fix this to delay stream removal until after all dependencies are removed.

WDYT?

**Comment by Scottmitch** [comment]
> In all the inbound frame handler methods, we could more-or-less ignore frames for streams that have been closed (i.e. existing-and-closed or the streamId <= lastKnownStream created by either Endpoint).

+1. Yes seems reasonable to me.

> We could be more picky and maintain some history of streamIds for which we've sent RST_STREAM in either direction and trigger errors if an Endpoint sends a (disallowed) frame after it had already sent RST_STREAM

Perhaps we don't maintain history by default....but it may be worth it to provide an interface which supports this?  Kinda like we have today....humm

> The only caveat to this (that I can think of ATM) would be the priority tree for outbound flow control.

Yes this may be problematic.  Lets handle this as a separate issue.  I think we need to think about this because `WINDOW_UPDATES` are only accepted for a `short period after a DATA or HEADERS frame containing an END_STREAM flag is sent` while PRIORITY frames can be received indefinitely.  I'm wondering if some kind of deadlock may occur if we keep something in the priority tree and the window can not be adjusted....

**Comment by nmittler** [comment]
> I'm wondering if some kind of deadlock may occur if we keep something in the priority tree and the window cannot be adjusted....

It seems reasonable to me that we always except WINDOW_UPDATES for streams that are still in the priority tree.  Assume stream A is closed and has children B and C. If we receive a `WINDOW_UPDATE` for A, I would expect it to still impact the data written for B and C. After  B and C are closed (and removed), we can then remove A. From that point forward we ignore WINDOW_UPDATES for stream A.

**Comment by Scottmitch** [comment]
> It seems reasonable to me that we always except WINDOW_UPDATES for streams that are still in the priority tree.

I agree.  However the specification may not allow sending of WINDOW_UPDATES in these states.  Just need to think about this I guess....

**Comment by Scottmitch** [comment]
Created new issue for the priority tree management https://github.com/netty/netty/issues/3518.

**Comment by buchgr** [comment]
ahh there is already an issue for this, I was thinking about opening one myself this afternoon :-). +1

**Comment by Scottmitch** [comment]
@nmittler - Is this still on your plate?

**Comment by nmittler** [comment]
@Scottmitch I'm was waiting for the dust to settle on some of the other changes to flow control.  I think everything is in now so I can take a look ... or were you looking to take this on?

**Comment by Scottmitch** [comment]
@nmittler - Nope I haven't touched this.

**Comment by Scottmitch** [comment]
@nmittler - Do you want https://github.com/netty/netty/pull/3543 to be pulled in before this?

**Comment by nmittler** [comment]
@Scottmitch yeah I think that makes sense.

**Comment by Scottmitch** [comment]
@nmittler - #3543 has been cherry-picked...you should now be unblocked.

**Comment by buchgr** [comment]
Nice to see this being fixed! We have seen some very large object allocations (several megabytes per object) due to the removal policy i.e. when opening and closing ~20,000 streams per second.

**Comment by Scottmitch** [comment]
@buchgr - Cool to see numbers and that you are pushing the codec hard...not cool that they appear to be less then optimal :)  I think we are in a better place to tackle this and I hope it simplifies things and helps with the number too.  Have you tried using the ImmediateRemovalPolicy instead?  Do you see any other areas for improvement that we are not currently working on?

**Comment by Scottmitch** [comment]
@nmittler - A few PRs have been cherry-picked since we last visited this.  Let me know if this is still blocked, or you are too busy to handle this.  I would like to get this in before https://github.com/netty/netty/issues/3501.

**Comment by buchgr** [comment]
@Scottmitch I talked to @nmittler and I ll submit a PR for this.
